### PR TITLE
7451 dvobject constraint, storageidentifier

### LINKED ISSUE #7451: Add a unique DvObjects/Storageidentifier constraint to legacy databases
Opening a new issue for the unfinished work in #6522.
In that issue the following constraint ("storageidentifiers should be unique within parent datasets") was added to DvObject.java:
```
uniqueConstraints = {...,@UniqueConstraint(columnNames = {"owner_id,storageidentifier"})}
``` 
The plan was to add this constraint to the existing databases using our normal flyway script method. We scrapped that plan once we realized that there's a certain number of legacy *harvested* datafiles in our database - and likely others - that have remote urls stored in the storageidentifier field; some of which are not unique within datasets.
It would be safe to actually add this constraint to older dbs such as ours.
These urls are not being used for any known purpose currently. We still don't want to lose them completely (?) - but we should think of a different place to move them to. 
Then we could reuse the flyway script originally created for the issue above: 
https://github.com/IQSS/dataverse/commit/cd5cf39f70bb17bac9cc4f9091cbbb6951f76ca8

### PR DESCRIPTION
**Pull request by landreev**
**What this PR does / why we need it**:

The constraint in question makes it impossible to create datafiles with the same storageidentifier, within the same dataset. (This translates into both datafiles pointing to the same physical file in the dataset directory on the filesystem, or the identical S3 object with the same pseudo-folder prefix). 
We still don't know of a way to reproduce the issue, but we know that it has happened in the past, where duplicate dvobjects/datafiles were created for uploaded files, for real users. This has disastrous consequences potentially (the user notices the duplicate, and attempts to remove it while it's still in draft; the physical file gets removed, leaving the other datafile pointing to the physical location that no longer exists!)
This *has been partially addressed* in v4.20, with the constraint added to the class file. This means that any database created since then has it enabled. However, we still haven't released a flyway script to force the constraint on older databases. This was because we realized that some legacy databases, notably ours, may have existing storageidentifiers - specifically, in harvested datafiles - violating the constraint. This PR addresses that. 

**Which issue(s) this PR closes**:

Closes #7451

**Special notes for your reviewer**:

Going into this, I was hoping to be able to address it all transparently, from a Flyway script, without having to ask the installations to do anything. I concluded that I could not achieve that. 
Back when we anticipated adding the constraint in v4.20, we asked the remote installations to run a diagnostic script checking for duplicates in their existing local (non-harvested) files, and helped those who have reported positives. However, this was a year ago, so we should assume that any pre-existing database may have gotten new instances of this issue since then. So we have no choice but to ask everybody to run the check again. 

One unpleasant discovery I made was that we, in our production, have had another case of this duplicated save since then, while already on v4.20. It happened in a more benign/less dangerous way - only the extra copies of DvObjects/DataFiles got created, *without any Filemetadata objects referencing them* (!?); so they were essentially invisible/practically inaccessible to anyone (wasn't immediately detected for that reason). But nevertheless, it did happen. 

When local, "real" files are concerned, we have to do it this way - asking to run diagnostics, then working with them to resolve any issues. It's an issue of data integrity, and I don't think we can create any automated solution for fixing it that would be reliable enough. There are too many different scenarios - is it published, is the physical file still there, are there any download counts associated with both dvobjects that they want to preserve, etc. 

So, the PR supplies another iteration of the diagnostics script, that we will want everybody to run before the next release (5.5, or whatever it is). This version will check all the harvested files as well. Unlike "real" local files, these are empty metadata shells that are only used for indexing  and generally considered disposable. So if any duplicates are found there, the script will just resolve it automatically in a straightforward/crude manner. 

Having said all that, if anyone has any idea of how to handle this in a more elegant manner, I will absolutely welcome that. 

**Suggestions on how to test this**:

Running the script on a real life db (and none is more real or messier that a copy of our own prod. db), then being able to deploy without Flyway barfing, is an ultimate test.
(edit: And after that it would also be a good idea to confirm that the constraint has been enabled. Maybe by attempting to set the storageidentifier of a datafile to something that's already in the database; which of course should be rejected)

**Does this PR introduce a user interface change? If mockups are available, please link/include them here**:

no

**Is there a release notes update needed for this change?**:

Yes, a *pre-release* note. 

**Additional documentation**:

### CODE REVIEW

File: src/main/resources/db/migration/V5.4.0.1__7451-unique-constraint-storageidentifier.sql [hunk 1]
```diff
@@ -0,0 +1 @@
+ALTER TABLE dvobject ADD CONSTRAINT unq_dvobject_storageidentifier UNIQUE(owner_id, storageidentifier);
```
[the remark this answers is not in the document]
Comment (qqmyers) [reply src/main/resources/db/migration/V5.4.0.1__7451-unique-constraint-storageidentifier.sql:1]: Ah - just read the comments...

[the remark this answers is not in the document]
Comment (landreev) [reply src/main/resources/db/migration/V5.4.0.1__7451-unique-constraint-storageidentifier.sql:1]: (and I only saw your first question initially, for some reason, not the second comment)

**Comment by landreev** [comment]
(I'm seeing that the last jenkins run bombed. I just re-synced the branch with develop, so this should trigger another test run; if that fails again, I'll investigate)

**Review by scolapasta — APPROVED** [review]

**Comment by kcondon** [comment]
@landreev Maybe add a note to instructions making it clearer to some folks like me:
Before attempting to deploy this release, which enforces a constraint, please run this script to detect any conflict, otherwise your deployment may fail.

**Comment by djbrooke** [comment]
Thx @landreev for adding the release note! I understand it and may tweak it during the release process when I'm bundling everything together.

**Pull request merged**